core/test/log/solver_debug.cpp

```diff
@@ -0,0 +1,177 @@
+           << std::setw(default_column_width) << T{4.0} << '\n'
+           << std::setw(default_column_width) << 1
+           << std::setw(default_column_width) << T{0.0}
+           << std::setw(default_column_width) << T{4.0}
+           << std::setw(default_column_width) << T{0.0}
+           << std::setw(default_column_width) << T{4.0}
+           << std::setw(default_column_width) << T{0.0} << '\n';
+    std::stringstream ss;
+    this->solver->add_logger(gko::log::SolverDebug::create_scalar_table(ss));
+
+    this->solver->apply(this->in, this->out);
+
+    ASSERT_EQ(ss.str(), ref_ss.str());
+}
+
+
+TYPED_TEST(SolverDebug, CsvWorks)
+{
+    using T = TypeParam;
+    std::stringstream ref_ss;
+    auto dynamic_type = gko::name_demangling::get_dynamic_type(*this->solver);
+    ref_ss << dynamic_type << "::apply(" << this->in.get() << ','
+           << this->out.get() << ") of dimensions " << this->solver->get_size()
+           << " and " << this->in->get_size()[1] << " rhs\n";
+    ref_ss << "Iteration,alpha,beta,prev_rho,rho,implicit_sq_residual_norm"
+           << '\n';
+    ref_ss << 0 << ',' << T{0.0} << ',' << T{0.0} << ',' << T{1.0} << ','
+           << T{4.0} << ',' << T{4.0} << '\n'
+           << 1 << ',' << T{0.0} << ',' << T{4.0} << ',' << T{0.0} << ','
+           << T{4.0} << ',' << T{0.0} << '\n';
+    std::stringstream ss;
+    this->solver->add_logger(gko::log::SolverDebug::create_scalar_csv(ss));
+
+    this->solver->apply(this->in, this->out);
+
+    ASSERT_EQ(ss.str(), ref_ss.str());
+}
+
+
+TYPED_TEST(SolverDebug, StorageWorks)
+{
+    using T = TypeParam;
+    using Dense = typename TestFixture::Dense;
+    auto orig_out = this->out->clone();
+    auto init_residual = gko::initialize<Dense>({T{-2.0}}, this->ref);
+    std::vector<std::pair<std::string, Dense*>> files{
+        {"solver_debug_test_0_alpha", this->zero.get()},
+        {"solver_debug_test_0_beta", nullptr},
+        {"solver_debug_test_0_implicit_sq_residual_norm", orig_out.get()},
+        {"solver_debug_test_0_minus_one", nullptr},
+        {"solver_debug_test_0_one", nullptr},
+        {"solver_debug_test_0_p", nullptr},
+        {"solver_debug_test_0_prev_rho", nullptr},
+        {"solver_debug_test_0_q", nullptr},
+        {"solver_debug_test_0_r", nullptr},
+        {"solver_debug_test_0_residual", init_residual.get()},
+        {"solver_debug_test_0_rho", nullptr},
+        {"solver_debug_test_0_solution", orig_out.get()},
+        {"solver_debug_test_0_z", nullptr},
+        {"solver_debug_test_1_alpha", nullptr},
+        {"solver_debug_test_1_beta", nullptr},
+        {"solver_debug_test_1_implicit_sq_residual_norm", this->zero.get()},
+        {"solver_debug_test_1_minus_one", nullptr},
+        {"solver_debug_test_1_one", nullptr},
+        {"solver_debug_test_1_p", nullptr},
+        {"solver_debug_test_1_prev_rho", nullptr},
+        {"solver_debug_test_1_q", nullptr},
+        {"solver_debug_test_1_r", nullptr},
```

Comment:
why are these (and many others) nullptrs? They should be filled after one iteration.

Reply:
I didn't want to check each individual file, only one from each "category" (workspace vector, each argument to `on_iteration`)

Reply:
That's fair. Maybe still check if the file exists?

Reply:
Good point, I only delete them right now